File: iconics-java/src/main/java/com/open/iconics/typeface/GenericFont.java
package com.open.iconics.typeface;

import android.content.Context;
import android.graphics.Typeface;

import java.util.Collection;
import java.util.HashMap;

public class GenericFont implements ITypeface {
    private String mFontName;
    private String mAuthor;
    private String mMappingPrefix;
    private String mFontFile;

    private Typeface typeface = null;
    private HashMap<String, Character> mChars = new HashMap<>();

    protected GenericFont() {
    }

    public GenericFont(String mappingPrefix, String fontFile) {
        this("GenericFont", "GenericAuthor", mappingPrefix, fontFile);
    }

    public GenericFont(String fontName, String author, String mappingPrefix, String fontFile) {
//        if (mappingPrefix.length() != 3) {
//            throw new IllegalArgumentException("MappingPrefix must be 3 char long");
//        }
        this.mFontName = fontName;
        this.mAuthor = author;
        this.mMappingPrefix = mappingPrefix;
        this.mFontFile = fontFile;
    }

    public void registerIcon(String name, char aChar) {
        mChars.put(mMappingPrefix + "_" + name, aChar);
    }

    @Override
    public IIcon getIcon(String key) {
        return new Icon(mChars.get(key)).withTypeface(this);
    }

    @Override
    public HashMap<String, Character> getCharacters() {
        return new HashMap<>();
    }

    @Override
    public String getMappingPrefix() {
        return mMappingPrefix;
    }

    @Override
    public String getFontName() {
        return mFontName;
    }

    @Override
    public String getVersion() {
        return "1.0.0";
    }

    @Override
    public int getIconCount() {
        return mChars.size();
    }

    @Override
    public Collection<String> getIcons() {
        return mChars.keySet();
    }

    @Override
    public String getAuthor() {
        return mAuthor;
    }

    @Override
    public String getUrl() {
        return "";
    }

    @Override
    public String getDescription() {
        return "";
    }

    @Override
    public String getLicense() {
        return "";
    }

    @Override
    public String getLicenseUrl() {
        return "";
    }

    @Override
    public Typeface getTypeface(Context context) {
        if (typeface == null) {
            try {
                typeface = Typeface.createFromAsset(context.getAssets(), mFontFile);
            } catch (Exception e) {
                return null;
            }
        }
        return typeface;
    }

    public class Icon implements IIcon {

        private String mName;
        private char aChar;
        private ITypeface mTypeface;

        public Icon(char c) {
            this.aChar = c;
        }

        public Icon(String name, char c) {
            this.mName = name;
            this.aChar = c;
        }

        public Icon withTypeface(ITypeface typeface) {
            this.mTypeface = typeface;
            return this;
        }

        @Override
        public String getFormattedName() {
            return "{" + getName() + "}";
        }

        @Override
        public String getName() {
            if (mName != null) {
                return mName;
            }
            return String.valueOf(aChar);
        }

        @Override
        public char getCharacter() {
            return aChar;
        }

        @Override
        public ITypeface getTypeface() {
            if (mTypeface != null) {
                return mTypeface;
            } else {
                return GenericFont.this;
            }
        }
    }
}

File: iconics-kotlin/src/main/java/com/open/iconics/typeface/GenericFont.kt
package com.open.iconics.typeface

import android.graphics.Typeface
import androidx.annotation.FontRes
import com.open.iconics.Iconics
import com.open.iconics.utils.IconicsPreconditions
import java.util.LinkedList

open class GenericFont : ITypeface {
    override val fontName: String

    override val author: String

    override val mappingPrefix: String

    override val fontRes: Int

    private val fontFile: String

    private val chars = HashMap<String, Char>()

    override val rawTypeface: Typeface
        get() = try {
            Typeface.createFromAsset(Iconics.applicationContext.assets, fontFile)
        } catch (ignored: Exception) {
            super.rawTypeface
        }

    override val characters: Map<String, Char>
        get() = chars

    override val version: String
        get() = "1.0.0"

    override val iconCount: Int
        get() = characters.size

    override val icons: List<String>
        get() = characters.keys.toCollection(LinkedList())

    override val url: String
        get() = ""

    override val description: String
        get() = ""

    override val license: String
        get() = ""

    override val licenseUrl: String
        get() = ""

    protected constructor() : this("GenericFont", "GenericAuthor", "", "")

    constructor(
        mappingPrefix: String,
        fontFile: String
    ) : this("GenericFont", "GenericAuthor", mappingPrefix, fontFile)

    constructor(
        mappingPrefix: String,
        @FontRes fontRes: Int
    ) : this("GenericFont", "GenericAuthor", mappingPrefix, fontRes)

    @Suppress("LeakingThis")
    constructor(fontName: String, author: String, mappingPrefix: String, fontFile: String) {
        IconicsPreconditions.checkMappingPrefix(mappingPrefix)
        this.fontName = fontName
        this.author = author
        this.mappingPrefix = mappingPrefix
        this.fontFile = fontFile
        this.fontRes = -1
    }

    @Suppress("LeakingThis")
    constructor(fontName: String, author: String, mappingPrefix: String, @FontRes fontRes: Int) {
        IconicsPreconditions.checkMappingPrefix(mappingPrefix)
        this.fontName = fontName
        this.author = author
        this.mappingPrefix = mappingPrefix
        this.fontFile = ""
        this.fontRes = fontRes
    }

    fun registerIcon(name: String, char: Char) {
        chars[mappingPrefix + "_" + name] = char
    }

    override fun getIcon(key: String): IIcon = Icon(key, chars.getValue(key)).withTypeface(this)

    inner class Icon : IIcon {
        private val customName: String?
        private var customTypeface: ITypeface? = null

        constructor(c: Char) {
            this.customName = null
            this.character = c
        }

        constructor(name: String, c: Char) {
            this.customName = name
            this.character = c
        }

        override val character: Char

        override val name: String
            get() = customName ?: character.toString()

        override val typeface: ITypeface
            get() = customTypeface ?: this@GenericFont

        fun withTypeface(typeface: ITypeface?): Icon {
            this.customTypeface = typeface
            return this
        }
    }
}
